xskillscore/core/contingency.py

```diff
@@ -0,0 +1,740 @@
+    """
+    header = f'<xskillscore.{type(self).__name__}>\n'
+    summary = header + '\n'.join(str(self.table).split('\n')[1:]) + '\n'
+    return summary
+
+
+class Contingency:
+    """Class for contingency based skill scores
+
+        Parameters
+        ----------
+        observations : xarray.Dataset or xarray.DataArray
+            Labeled array(s) over which to apply the function.
+        forecasts : xarray.Dataset or xarray.DataArray
+            Labeled array(s) over which to apply the function.
+        observation_category_edges : array_like
+            Bin edges for categorising observations
+        forecast_category_edges : array_like
+            Bin edges for categorising forecasts
+        dim : str, list
+            The dimension(s) over which to compute the contingency table
+
+        Returns
+        -------
+        xskillscore.Contingency
+
+        Examples
+        --------
+        >>> a = xr.DataArray(np.random.normal(size=(3,3)),
+        ...                  coords=[('x', np.arange(3)), ('y', np.arange(3))]).to_dataset(name='test1')
+        >>> b = xr.DataArray(np.random.normal(size=(3,3)),
+        ...                  coords=[('x', np.arange(3)), ('y', np.arange(3))]).to_dataset(name='test1')
+        >>> a['test2'] = xr.DataArray(np.random.normal(size=(3,3)),
+        ...                           coords=[('x', np.arange(3)), ('y', np.arange(3))])
+        >>> b['test2'] = xr.DataArray(np.random.normal(size=(3,3)),
+        ...                           coords=[('x', np.arange(3)), ('y', np.arange(3))])
+        >>> category_edges_a = np.linspace(-2,2,5)
+        >>> category_edges_b = np.linspace(-3,3,5)
+        >>> Contingency(a, b, category_edges_a, category_edges_b, dim=['x','y'])
+        <xskillscore.Contingency>
+        Dimensions:                       (forecasts_category: 4, observations_category: 4)
+        Coordinates:
+            observations_category_bounds  (observations_category) <U12 '(-2.0, -1.0]'...
+            forecasts_category_bounds     (forecasts_category) <U12 '(-3.0, -1.5]' .....
+          * observations_category         (observations_category) int64 1 2 3 4
+          * forecasts_category            (forecasts_category) int64 1 2 3 4
+        Data variables:
+            test2                         (observations_category, forecasts_category) int64 ...
+            test1                         (observations_category, forecasts_category) int64 ...
+
+        Notes
+        -----
+        See http://www.cawcr.gov.au/projects/verification/
+    """
+
+    def __init__(
+        self,
+        observations,
+        forecasts,
+        observation_category_edges,
+        forecast_category_edges,
+        dim,
+    ):
+        self.observations = observations
+        self.forecasts = forecasts
+        self.observation_category_edges = observation_category_edges
+        self.forecast_category_edges = forecast_category_edges
+        self.dichotomous = (
```

Comment:
I'm still learning best practices for object-oriented development. I'm not sure if it's worth it here, but generally you want to go with getters/setters for "data encapsulation." I think essentially to make sure these primary features can't be modified accidentally (see https://www.geeksforgeeks.org/getter-and-setter-in-python/).

This would look like:

```python
def __init__(...):
    self._observations = observations
    self._forecasts = forecasts
    # and so on...

@property
def observations(self):
    return self._observations

@property
def forecasts(self):
    return self._forecasts
```

This is so someone can't do

```python
obj = Contingency(...)
obj.forecast = 3
```

I don't think you need to add the "setters" here since they shouldn't be able to modify these primary attributes after the object is instantiated.

Reply:
Also you might need to do `self._observations = observations.copy()` and so on. There's weird cases where if you don't copy it, you'll modify the original object with any modifications that happen by running the methods. It's unclear to me when this happens.

Reply:
Good advice - thanks! Changes made.